hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metrics/m3/M3MetricsReporter.java

```diff
@@ -0,0 +1,120 @@
+  private final ImmutableMap<String, String> tags;
+
+  public M3MetricsReporter(HoodieWriteConfig config, MetricRegistry registry) {
+    this.config = config;
+    this.registry = registry;
+
+    ImmutableMap.Builder tagBuilder = new ImmutableMap.Builder<>();
+    tagBuilder.putAll(parseOptionalTags(config.getM3Tags()));
+    tagBuilder.put("service", config.getM3Service());
+    tagBuilder.put("env", config.getM3Env());
+    this.tags = tagBuilder.build();
+    LOG.info(String.format("Building M3 Reporter with M3 tags mapping: %s", tags));
+  }
+
+  private static Map parseOptionalTags(String tagValueString) {
+    Map parsedTags = new HashMap();
+    if (!tagValueString.isEmpty()) {
+      Arrays.stream(tagValueString.split(",")).forEach((tagValuePair) -> {
+        String[] parsedTagValuePair = Arrays.stream(tagValuePair.split("="))
+            .map((tagOrValue) -> tagOrValue.trim()).filter((tagOrValue) -> !tagOrValue.isEmpty())
+            .toArray(String[]::new);
+        if (parsedTagValuePair.length != 2) {
+          throw new RuntimeException(String.format(
+              "M3 Reporter tags cannot be initialized with tags [%s] due to not being in format `tag=value, . . .`.",
+              tagValuePair));
+        }
+        parsedTags.put(parsedTagValuePair[0], parsedTagValuePair[1]);
+      });
+    }
+    return parsedTags;
+  }
+
+  @Override
+  public void start() {}
+
+  @Override
+  public void report() {
+    /*
+      Although com.uber.m3.tally.Scope supports automatically submitting metrics in an interval
+      via a background task, it does not seem to support
+      - an API for explicitly flushing/emitting all metrics
+      - Taking in an external com.codahale.metrics.MetricRegistry metrics registry and automatically
+      adding any new counters/gauges whenever they are added to the registry
+      Due to this, this implementation emits metrics by creating a Scope, adding all metrics from
+      the HUDI metircs registry as counters/gauges to the scope, and then closing the Scope. Since
+      closing this Scope will implicitly flush all M3 metrics, the reporting intervals
+      are configured to be Integer.MAX_VALUE.
+     */
+    synchronized (this) {
+      try (Scope scope = new RootScopeBuilder()
+          .reporter(new M3Reporter.Builder(
+              new InetSocketAddress(config.getM3ServerHost(), config.getM3ServerPort()))
+              .includeHost(true).commonTags(tags)
+              .build())
+          .reportEvery(Duration.ofSeconds(Integer.MAX_VALUE))
+          .tagged(tags)) {
+
+        M3ScopeReporterAdaptor scopeReporter = new M3ScopeReporterAdaptor(registry, scope);
+        scopeReporter.start(Integer.MAX_VALUE, TimeUnit.SECONDS);
+        scopeReporter.report();
+        scopeReporter.stop();
+      } catch (Exception e) {
+        LOG.error(String.format("Error reporting metrics to M3: %s", e));
+      }
+    }
+  }
+
+  @Override
+  public void stop() {}
+}
+
```

Comment:
So there is nothing to initialize and release.

Reply:
Yes, resources are only created and released during report() call.